### PR TITLE
Kube-dns configmap translation to CoreDNS

### LINKED ISSUE #662: Kube-dns configmap translation to CoreDNS
FEATURE REQUEST

Currently, when we install CoreDNS instead of kube-dns via `kubeadm upgrade`, we do not carry forward the configuration of kube-dns.
We need to translate and transfer the configmap and deployment settings of kube-dns into a CoreDNS Corefile that will implement the same features to ensure a smooth transition when upgrading to CoreDNS.

### PR DESCRIPTION
**Pull request by rajansandeep**
**What this PR does / why we need it**:
Translation of kube-dns ConfigMap (StubDomains and UpstreamNameserver) to equivalent Proxy in CoreDNS.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#662

**Special notes for your reviewer**:

Currently, the translation is done as follows:
Example ConfigMap of kube-dns.
```
apiVersion: v1
data:
  federations: |
    {"foo" : "foo.feddomain.com"}
  stubDomains: |
    {"abc.com" : ["1.2.3.4"], "my.cluster.local" : ["2.3.4.5"]}
  upstreamNameservers: |
    ["8.8.8.8", "8.8.4.4"]
kind: ConfigMap
metadata:
  creationTimestamp: 2018-01-22T20:21:56Z
  name: kube-dns
  namespace: kube-system
```

CoreDNS Corefile after translation.

```
   .:53 {
        errors
        health
        kubernetes cluster.local  in-addr.arpa ip6.arpa {
           upstream  8.8.8.8 8.8.4.4
           pods insecure
           fallthrough in-addr.arpa ip6.arpa
        }
       federation cluster.local {
           foo foo.feddomain.com
        }
        prometheus :9153
        proxy .  8.8.8.8 8.8.4.4
        cache 30
    }
    abc.com:53 {
        errors
        cache 30
        proxy . 1.2.3.4
    }
    my.cluster.local:53 {
        errors
        cache 30
        proxy . 2.3.4.5
    }
```

**Release note**:

```release-note
Kubeadm: CoreDNS supports migration of the kube-dns configuration to CoreDNS configuration when upgrading the service discovery from kube-dns to CoreDNS as part of Beta. 
```

### CODE REVIEW

**Comment by fturib** [comment]
@bowei , @luxas : could you provide your opinion on this translation of kube-dns configuration into a coredns configuration ?

We support here the options defined in the ConfigMap of kube-dns, not the options on the command line of kube-dns's containers. 
We think we are addressing the change that user's can do 'easily' within k8s environment, I mean the configuration advertised in documentations for kube-dns.

**Comment by kargakis** [comment]
@kubernetes/sig-network-pr-reviews who is responsible for the coredns integration? Can they have an eye on this?

**Comment by rajansandeep** [comment]
@errordeveloper @MrHohn PTAL.

**Comment by fturib** [comment]
/assign @bowei 

Hi, Bowei, thanks for reviewing

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (bowei) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go]: is this just a newline character? could we use "\n"

**Review by bowei — COMMENTED** [review]
what about federation? that is also part of the configmap

**Comment by rajansandeep** [comment]
@bowei I added the federation translation.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: kill this newline

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: and this

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: Is it worth it to pull a kubelet library for a single constant?

Comment (errordeveloper) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: It looks like a simple package and won't add too much dead code, but the value of constant is very generic – `/etc/resolv.conf`, the chances it changes are small and we probably have it as literal string in many other places too.

**Review by errordeveloper — COMMENTED** [review]
I'd much rather see us using proper codec here, and not muck around with strings. Have you considered using [`github.com/mholt/caddy/caddyfile`](https://godoc.org/github.com/mholt/caddy/caddyfile)?

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: Could you please comment block above this line to outline why this here?

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Done.

**Comment by rajansandeep** [comment]
@errordeveloper I have made changes to use `github.com/mholt/caddy/caddyfile` as suggested.

**Comment by rajansandeep** [comment]
Seems like Caddy package isn't available.

`build aborted: no such package 'vendor/github.com/mholt/caddy/caddyfile': BUILD file not found on package path`

**Review by timothysc — APPROVED** [review]
/approve  - I'll let others wrangle on the lgtm details.

**Comment by rajansandeep** [comment]
@errordeveloper How do I vendor the caddy package?

**Comment by errordeveloper** [comment]
Heve you looked at
https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md
?

**Comment by errordeveloper** [comment]
@rajansandeep have you looked at [`/community/contributors/devel/godep.md`](https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md)?

**Comment by rajansandeep** [comment]
@errordeveloper thank you for showing me the steps to vendor. PTAL to see if all looks ok.
Also, should this PR be tagged with the 1.10 milestone?

**Comment by fturib** [comment]
@bowei : can you have a second look after Sandeep added the Federation piece translation ? Thanks!

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (fturib) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: all these lines 364-376 is duplicated from Federation to StubDomain part.
Can you make a function structureToCorefileStanza(data map[string]interface{}) (string, err) that just take any data and translate properly into stanza

Comment (fturib) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: unless it is not the same ? (because one is a sub-part ?)

**Comment by fturib** [comment]
@bowei, @MrHohn  : gentle reminder here, as the code slush is today !. Thanks for advises or good to go mark.

**Comment by rajansandeep** [comment]
/assign @monopole 
For the caddy vendor part.

**Comment by fturib** [comment]
@timothysc : can you flag this PR for v1.10. ?
The corresponding issue is already, but in kubernetes/kubeadm.
I would like this PR get some visibility to have some traction for review.
It is needed for CoreDNS going as BETA for v1.10.
Thanks!!

kubernetes/kubeadm#662

**Comment by fturib** [comment]
/assign @thockin 

@thockin : Hi Tim, this PR is needed for CoreDNS to go BETA for v1.10, as expected in plan of SIG Network. 
Can you flag this PR for v1.10 ? 
Thanks!

the corresponding issue is flagged already : kubernetes/kubeadm#662
the CrieDNS feature is flagged too : kubernetes/features#427

**Comment by fturib** [comment]
As of today, we have the update of last version of CoreDNS 1.0.6.
@rajansandeep : can you update this PR with the latest manifest and adapt the subdomain translation with the new Upstream option available.

Thanks!

**Comment by fturib** [comment]
## This PR contains all changes needed for CoreDNS going BETA feature for kubeadm.
- the translation of the Kube-dns configmap configuration at upgrade time
- bump-up to last release of coreDNS 1.0.6
- flag as BETA feature in kubeadm


## Thanks for review

**Comment by fturib** [comment]
/assign @cblecker
@cblecker : can you approve the vendor part of this PR ? or assign to someone that can review/approve ?

Thanks!!!

**Comment by cblecker** [comment]
As this is a new dependency, @thockin needs to review.
/unassign
/assign @thockin

**Comment by thockin** [comment]
I don't have context to review this.  Can someone who knows kubeadm well please take a look?

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: This doesn't capture the return value, what is it for?

Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Yes, that line is redundant. Shall remove.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: This seems too crude. Why not use BASE64 or double-encode as JSON?

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: Relative indentation assumptions is a terrible thing to do, why?

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [inline cmd/kubeadm/app/phases/addons/dns/dns.go]: This looks like it is tinny bit different from the above, it's very error-prone.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: The reason I had to use the string replacer is because of the difference in formatting in JSON and YAML.
Since the manifest is parsed as a yaml, it errors out due to tabs instead of spaces and so on….

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: The indentation isn’t strictly necessary for the conversion, but I insisted on it because after the upgrade is done and the user wants to modify the configmap, it ensures that the configmap is in a readable format and this makes it easier for the user to interact.

**Comment by rajansandeep** [comment]
@errordeveloper @timothysc @thockin @luxas: Are there any other labels needed to be added to this PR like "approved-for-milestone"?

**Comment by thockin** [comment]
Assign back to me once it is LGTM'ed and ready for approval.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: I am not entirely sure this is an essential kind of optimisation. Please add a comment to clarify the intention.

File: cmd/kubeadm/app/phases/addons/dns/dns.go
Comment (errordeveloper) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: I understand why you ended up having to do this, but this is a very crude method. Have you tried double encoding it?

Comment (fturib) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: I am not sure I understand fully, but the double encoding would avoid the replacement of \" .. not really the formatting part - I mean we still need the replacement of \t, \n with space etc ...

Looks like there is no nice way to use the original codec (caddyfile). Because we then endup with incompatibility between JSON and YAML, and then we have anyway the formatting concern because that config can be edited directly as kubectl cmd. (we want it to be kind-of human readable format).

So what is the best to do here, without over complicating the code - that is, at the end, used only once. ?

Go back to initial code with simple string translation ?

@errordeveloper : what do you think ?

Comment (kargakis) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Why not use [caddyfile](https://godoc.org/github.com/mholt/caddy/caddyfile)? It's the official package for parsing caddyfiles and it's self-contained (only stdlib code).

Comment (kargakis) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: Oh, it's already used above. Can't you use https://github.com/kubernetes/kubernetes/tree/master/vendor/github.com/ghodss/yaml to go from JSON to YAML?

Comment (kargakis) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: yaml.JSONToYAML

Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/dns.go]: I cannot because the output is in the form of caddyfile text, which will be parsed into the existing Corefile in the configmap.
(The variable name outputJSON is misleading, I'll fix that)

**Comment by rajansandeep** [comment]
@kargakis @errordeveloper PTAL.

**Comment by rajansandeep** [comment]
@errordeveloper Please review the PR and let me know if any changes are required for LGTM tag. Thanks!

**Comment by errordeveloper** [comment]
/lgmt

This looks much better now! Broadly, I still find it is a bit too much of a voodoo to be honest, I think there is a need for a proper API somewhere between the layers that this code glues together, but I understand that it would be too much of an ask for the task at hand, perhaps something to think about. Maybe CoreDNS needs an operator and/or CRD, or a plugin, not sure exactly, and I only have very limited context of how all the pieces integrate, so please evaluate this suggestion within the context of your own.

**Comment by fturib** [comment]
@errordeveloper : thank your for go ahead and the advice.
Could you provide the "lgtm" with the right syntax ? .. in order to have the whole approval ?

Thanks!

**Comment by rajansandeep** [comment]
/assign @thockin 
Assigning as PR has lgtm tag.

**Comment by fturib** [comment]
/assign thockin

@thockin : Hi Tim, we have now approval and lgtm for the kubeadm part.
We need now your approval for the vendorization part (see comment above from @cblecker).

The reasons of this add of "caddyfile" lib in vendors is that it was asked during this review to use the right codec to write the configuration file for CoreDNS, instead of a simple string manipulation inside the configmap. 
@rajansandeep did that translation, including "caddyfile" lib for the marshalling into Corefile for CoreDNS.

Thanks for approval

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/58828)

**Pull request merged**
